### PR TITLE
IoT Central 1.1-preview

### PR DESCRIPTION
**Pull request by lucadruda**
- [x] If introducing new functionality or modified behavior, are they backed by unit and integration tests?
- [x] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [x] Have **all** unit **and** integration tests passed locally? i.e. `pytest <project root> -vv`. Please provide evidence in the form of a screenshot showing a successful run of integration tests.
- [x] Have static checks passed using the .pylintrc and .flake8 rules? Look at the CI scripts for example usage.
- [x] Have you made an entry in HISTORY.rst which concisely explains your user-facing feature or change?

* Adds commands reflecting API v1.1-preview.
* Refactor all IoT Central list commands to align with CLI standards (breaking change)
* Includes header "x-ms-client-request-id" in each request.
* Add test option "--api-version" to support test against multiple api versions

Integration tests:
imageimage

### CODE REVIEW

**Comment by vilit1** [comment]
Have all comments from the previous pr been addressed?

**Comment by lucadruda** [comment]
This PR totally replaces #421. Previous comments have been addressed in this new PR.
Refactoring involved creation of model folders matching different api versions. Duplicated models (e.g. DeviceV1 does not differ from DeviceV2) are kept to allow devs to easily remove deprecated versions in the future and isolate modules.

**Review by yang-hai-feng — DISMISSED** [review]

File: azext_iot/central/commands_file_upload.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+from azext_iot.central.providers import CentralFileUploadProvider
+from azext_iot.central.models.enum import ApiVersion
+
+
+def get_fileupload(
+    cmd,
+    app_id: str,
+    token=None,
+    central_dns_suffix=CENTRAL_ENDPOINT,
+    api_version=ApiVersion.v2.value,
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/commands_file_upload.py:18]: sure. aligning all to "1.1-preview"

File: azext_iot/central/commands_job.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+from azext_iot.common import utility
+
+
+def get_job(
+    cmd,
+    app_id: str,
+    job_id: str,
+    token=None,
+    central_dns_suffix=CENTRAL_ENDPOINT,
+    api_version=ApiVersion.preview.value,
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/commands_job.py:21]: using 1.1-preview as default for all "preview" commands

File: azext_iot/central/providers/device_provider.py [hunk 3]
```diff
@@ -42,7 +46,7 @@ def get_device(
         self,
         device_id,
         central_dns_suffix=CENTRAL_ENDPOINT,
-    ) -> DeviceV1:
+    ) -> Union[DeviceV1, DeviceV2, DevicePreview]:
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/providers/device_provider.py:49]: do you mean revert to previous structure with multiple provider versions?

File: azext_iot/central/models/enum.py [hunk 4]
```diff
@@ -60,3 +60,4 @@ class ApiVersion(Enum):
 
     preview = "preview"
     v1 = "1.0"
+    v2 = "1.1-preview"
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/models/enum.py]: fixed as above

File: azext_iot/central/_help.py [hunk 5]
```diff
@@ -512,6 +562,195 @@ def _load_central_roles_help():
...
+
+      - name: Create a job with name and cancellation threshold configuration in percentage with no batch
+        text: >
+          az iot central job create
+          --app-id {appid}
+          --job-id {jobId}
+          --group-id {groupId}
+          --job-name {jobName}
+          --content {creationJSONPath}
+          --threshold {thresholdValue}
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/_help.py:682]: modified

[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/_help.py:682]: renamed

File: azext_iot/central/command_map.py [hunk 6]
```diff
@@ -131,6 +141,34 @@ def load_central_commands(self, _):
...
         cmd_group.command("list", "list_roles")
 
+    with self.command_group(
+        "iot central file-upload-configuration",
+        command_type=central_file_uploads_ops,
+        is_preview=True,
+    ) as cmd_group:
+        cmd_group.show_command("show", "get_fileupload")
+        cmd_group.show_command("delete", "delete_fileupload")
+        cmd_group.show_command("create", "create_fileupload")
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/command_map.py:151]: "Update" has never been available for all commands in the central CLI (not even in 1.0).
All other CLI commands in the extension have the "update" operation merged into "create" when existing Id is passed.

File: azext_iot/central/command_map.py [hunk 7]
```diff
@@ -131,6 +141,34 @@ def load_central_commands(self, _):
...
+
+    with self.command_group(
+        "iot central organization", command_type=central_orgs_ops, is_preview=True
+    ) as cmd_group:
+        cmd_group.show_command("show", "get_org")
+        cmd_group.command("list", "list_orgs")
+        cmd_group.command("create", "create_org")
+        cmd_group.command("delete", "delete_org")
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/command_map.py:159]: same as above

File: azext_iot/central/params.py [hunk 8]
```diff
@@ -169,6 +181,17 @@ def load_central_arguments(self, _):
...
         )
 
+    with self.argument_context("iot central device-group") as context:
+        context.argument(
+            "api_version",
+            options_list=["--api-version", "--av"],
+            choices=CaseInsensitiveList(
+                [ApiVersion.preview.value, ApiVersion.v2.value]
+            ),
+            default=ApiVersion.v2.value,
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/params.py]: Current default has been deprecated in API. What are the requirements in this case?

File: azext_iot/central/providers/job_provider.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+        return jobs
+
+    def get_job(
+        self,
+        job_id,
+        central_dns_suffix=CENTRAL_ENDPOINT,
+    ) -> Union[JobPreview, JobV2]:
+        # get or add to cache
+        job = self._jobs.get(job_id)
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/providers/job_provider.py:60]: This is all legacy behavior. All these things should be part of a different discussion and refactoring

File: azext_iot/central/models/v2/device.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        self.etag = device.get("etag")
+        self.id = device.get("id")
+        self.template = device.get("template")
+        self.provisioned = device.get("provisioned")
+        self.simulated = device.get("simulated")
+        self.organizations = device.get('organizations')
+        self.device_status = self._parse_device_status()
+
+    def _parse_device_status(self) -> DeviceStatus:
+        if not self.enabled:
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/models/v2/device.py]: could you elaborate more? this hasn't been changed

File: azext_iot/central/commands_organization.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+
+def get_org(
+    cmd,
+    app_id: str,
+    org_id: str,
+    token=None,
+    central_dns_suffix=CENTRAL_ENDPOINT,
+    api_version=ApiVersion.v2.value,
+):
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/commands_organization.py:20]: At this point of flow return types are always dict or list of dict. We didn't have return types at command level (only at services). Do you mean of adding return model type here or just "dict", List[dict] ?

File: azext_iot/central/providers/job_provider.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+            threshold_batch=threshold_batch,
+            batch=batch,
+            threshold=threshold,
+            token=self._token,
+            api_version=self._api_version,
+            central_dns_suffix=central_dns_suffix,
+        )
+
+        if not job:
+            raise CLIError("No job found with id: '{}'.".format(job_id))
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/providers/job_provider.py:186]: fixed

File: azext_iot/central/services/_utility.py [hunk 13]
```diff
@@ -49,3 +54,37 @@ def log_response_debug(response: Response, logger: logging.Logger):
     logger.debug("Response status code: {}".format(response.status_code))
     logger.debug("Response url: {}".format(response.url))
     logger.debug("Response headers: {}".format(response.headers))
+
+
+def get_object(data: dict, model: str, api_version):
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/services/_utility.py:59]: impossible to have a specific return type as this is a helper to dynamically return requested object based on parameters.
I'll put "object" as base for the class hierarchy

File: azext_iot/central/services/device.py [hunk 14]
```diff
@@ -219,39 +229,37 @@ def create_device(
...
-            "simulated": simulated,
-            "enabled": True,
-        }
-        if template:
-            payload["template"] = template
+    payload = {
+        "displayName": device_name,
+        "simulated": simulated,
+        "enabled": True,
+    }
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/services/device.py:236]: great catch!

File: azext_iot/central/services/job.py [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+        app_id: name of app (used for forming request URL)
+        token: (OPTIONAL) authorization token to fetch job details from IoTC.
+            MUST INCLUDE type (e.g. 'SharedAccessToken ...', 'Bearer ...')
+        central_dns_suffix: {centralDnsSuffixInPath} as found in docs
+
+    Returns:
+        job: dict
+    """
+
+    return _list_job(
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/services/job.py:293]: this is a helper to call GET apis returning a list. Get_Devices maps this: https://docs.microsoft.com/en-us/rest/api/iotcentral/1.1-previewdataplane/jobs/get-devices

**Review by yang-hai-feng — APPROVED** [review]

File: azext_iot/central/command_map.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/command_map.py:151]: Separate work item created

File: azext_iot/central/providers/job_provider.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/providers/job_provider.py:61]: Created work item

File: azext_iot/tests/central/test_iot_central_int.py [hunk 16]
```diff
@@ -278,3 +285,31 @@ def test_central_run_command_component(self):
 
         # check that run result and show result indeed match
         assert run_result["response"] == show_result["value"][0]["response"]
+
+    @pytest.mark.skipif(
+        not STORAGE_CSTRING or not STORAGE_CONTAINER,
+        reason="empty azext_iot_central_storage_cstring or azext_iot_central_storage_container env var",
+    )
+    def test_central_fileupload_methods_CRD(self):
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/tests/central/test_iot_central_int.py]: done

File: azext_iot/tests/central/test_iot_central_int.py [hunk 17]
```diff
@@ -278,3 +285,31 @@ def test_central_run_command_component(self):
...
+            TOKEN,
+            DNS_SUFFIX,
+        )
+        result = self.cmd(command).get_output_in_json()
+        assert result["connectionString"] == file_upload["connectionString"]
+        assert result["container"] == file_upload["container"]
+        self._delete_fileupload()
+
+    def test_central_organization_methods_CRD(self):
+        org = self._create_organization()
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/tests/central/test_iot_central_int.py]: currently there is an open bug on file upload so checks on failure are useless now.

**Review by yang-hai-feng — DISMISSED** [review]

File: azext_iot/central/providers/organization_provider.py [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+logger = get_logger(__name__)
+
+
+class CentralOrganizationProvider:
+    def __init__(self, cmd, app_id: str, api_version: str, token=None):
+        """
+        Provider for organizations APIs
+
+        Args:
+            cmd: command passed into az
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/providers/organization_provider.py:24]: fixed

File: azext_iot/tests/central/test_iot_central_int.py [hunk 19]
```diff
@@ -278,3 +285,31 @@ def test_central_run_command_component(self):
...
+    def test_central_fileupload_methods_CRD(self):
+        file_upload = self._create_fileupload()
+        self._wait_for_storage_configured()
+        command = self._appendOptionalArgsToCommand(
+            "iot central file-upload-configuration show -n {}".format(APP_ID),
+            TOKEN,
+            DNS_SUFFIX,
+        )
+        result = self.cmd(command).get_output_in_json()
+        assert result["connectionString"] == file_upload["connectionString"]
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/tests/central/test_iot_central_int.py:356]: current bug on API prevents to have better checks. will change once API get fixed

File: azext_iot/central/providers/job_provider.py [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+        content,
+        batch_percentage,
+        threshold_percentage,
+        threshold_batch,
+        batch,
+        threshold,
+        description,
+        central_dns_suffix=CENTRAL_ENDPOINT,
+    ):
+        if job_id in self._jobs:
```
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/providers/job_provider.py:166]: will work on cache in different task

File: azext_iot/central/services/_utility.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/services/_utility.py]: added types everywhere

File: azext_iot/central/commands_organization.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (lucadruda) [reply azext_iot/central/commands_organization.py]: added

**Review by yang-hai-feng — APPROVED** [review]

File: dev_requirements [hunk 21]
```diff
@@ -4,12 +4,14 @@ pytest-cov
...
 wheel==0.30.0
 pre-commit
 pylint
 flake8
 azure-storage-blob
 pytest-xdist
-pytest-rerunfailures
\ No newline at end of file
+pytest-rerunfailures
+tqdm
```
Comment (digimaun) [inline dev_requirements:16]: `tqdm` and `jsonschema` should be installed with the extension and are not `dev_requirements` items.

File: jobdata.json [hunk 22]
```diff
@@ -0,0 +1,12 @@
+[
```
Comment (digimaun) [inline jobdata.json:1]: Assuming the placement of this json file is not intended?

File: azext_iot/central/models/v1/role.py [hunk 23]
```diff
@@ -0,0 +1,12 @@
...
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+
+class Role:
+    def __init__(self, device: dict):
+        self.display_name = device.get("displayName")
+        self.id = device.get("id")
+        pass
```
Comment (digimaun) [inline azext_iot/central/models/v1/role.py:12]: Unneeded `pass`

File: azext_iot/central/models/v1_1_preview/device_group.py [hunk 24]
```diff
@@ -0,0 +1,13 @@
...
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+
+class DeviceGroup:
+    def __init__(self, group: dict):
+        self.display_name = group.get("displayName")
+        self.id = group.get("id")
+        self.organizations = group.get("organizations")
+        pass
```
Comment (digimaun) [inline azext_iot/central/models/v1_1_preview/device_group.py:13]: Unneeded `pass`

File: azext_iot/central/models/v1_1_preview/role.py [hunk 25]
```diff
@@ -0,0 +1,12 @@
...
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+
+class Role:
+    def __init__(self, device: dict):
+        self.display_name = device.get("displayName")
+        self.id = device.get("id")
+        pass
```
Comment (digimaun) [inline azext_iot/central/models/v1_1_preview/role.py:12]: Unneeded `pass`

File: azext_iot/central/params.py [hunk 26]
```diff
@@ -155,10 +160,17 @@ def load_central_arguments(self, _):
...
             "device_id",
             options_list=["--device-id", "-d"],
-            help="Provide a unique identifier for the device."
+            help="Unique identifier for the device."
             " A case-sensitive string (up to 128 characters long) of ASCII 7-bit alphanumeric characters plus"
             " certain special characters: - . + % _ # * ? ! ( ) , : = @ $ '",
         )
+        context.argument(
+            "organizations",
+            options_list=["--organizations", "--org"],
```
Comment (digimaun) [inline azext_iot/central/params.py:169]: `--organizations` is plural but `--org` is not?

File: azext_iot/central/params.py [hunk 27]
```diff
@@ -104,6 +103,12 @@ def load_central_arguments(self, _):
             " Specify an ID that you'll then use when modifying or deleting this token later via the CLI or API.",
         )
         context.argument("role", arg_type=role_type)
+        context.argument(
+            "org_id",
+            options_list=["--organization-id", "--orgid"],
```
Comment (digimaun) [inline azext_iot/central/params.py:108]: The short alias should be `--org-id` to meet the conventions of existing args here are examples from this PR

image

File: azext_iot/central/params.py [hunk 28]
```diff
@@ -214,12 +237,152 @@ def load_central_arguments(self, _):
...
+
+    with self.argument_context("iot central organization") as context:
+        context.argument(
+            "org_id",
+            options_list=["--org-id"],
+            help="Unique identifier for the organization.",
+        )
+        context.argument(
+            "api_version",
+            options_list=["--api-version", "--av"],
```
Comment (digimaun) [inline azext_iot/central/params.py:292]: Try and de-dupe the parameters as much as possible.

Assuming `--api-version` being defined multiple times for different command groups is due to inconsistent API version support in central i.e. ApiVersion.v1_1_preview and ApiVersion.preview support for `az iot central job` vs only ApiVersion.v1_1_preview support in `az iot central organization`.

File: azext_iot/central/models/preview/device_group.py [hunk 29]
```diff
@@ -5,8 +5,9 @@
 # --------------------------------------------------------------------------------------------
 
 
-class DeviceGroupPreview:
+class DeviceGroup:
     def __init__(self, group: dict):
         self.display_name = group.get("displayName")
         self.id = group.get("id")
+        self.organizations = group.get("organizations")
         pass
```
Comment (digimaun) [inline azext_iot/central/models/preview/device_group.py:13]: Unneeded `pass`

File: azext_iot/central/_help.py [hunk 30]
```diff
@@ -512,6 +562,201 @@ def _load_central_roles_help():
     """
 
 
+def _load_central_organizations_help():
+    helps[
+        "iot central organization"
```
Comment (digimaun) [inline azext_iot/central/_help.py:567]: Can `organization` be shortened to `org` for this command group, or is it important to spell it out?

File: azext_iot/central/_help.py [hunk 31]
```diff
@@ -512,6 +562,201 @@ def _load_central_roles_help():
...
+        "iot central job create"
+    ] = """
+    type: command
+    short-summary: Create and execute a job via its job definition.
+    examples:
+      - name: Create a job with name
+        text: >
+          az iot central job create
+          --app-id {appid}
+          --job-id {jobId}
```
Comment (digimaun) [inline azext_iot/central/_help.py:664]: I find it strange to have both `--job-id` and `--job-name` for the same command. Can `--job-name` be changed to `--display-name` or similar if that is what it is intended for.

File: azext_iot/central/_help.py [hunk 32]
```diff
@@ -477,12 +480,59 @@ def _load_central_device_groups_help():
     """
 
 
+def _load_central_file_upload_configuration_help():
+    helps[
+        "iot central file-upload-configuration"
```
Comment (digimaun) [inline azext_iot/central/_help.py:485]: Can this command group be shortened to [az] `iot central file-upload-config` ?

File: azext_iot/tests/central/__init__.py [hunk 33]
```diff
@@ -6,27 +6,53 @@
...
-from azext_iot.central.models.enum import Role
+from azext_iot.central.models.enum import Role, UserTypePreview, UserTypeV1, ApiVersion
 
 APP_ID = os.environ.get("azext_iot_central_app_id")
 APP_PRIMARY_KEY = os.environ.get("azext_iot_central_primarykey")
 APP_SCOPE_ID = os.environ.get("azext_iot_central_scope_id")
 DEVICE_ID = os.environ.get("azext_iot_central_device_id")
 TOKEN = os.environ.get("azext_iot_central_token")
 DNS_SUFFIX = os.environ.get("azext_iot_central_dns_suffix")
+STORAGE_CSTRING = os.environ.get("azext_iot_central_storage_cstring")
```
Comment (digimaun) [inline azext_iot/tests/central/__init__.py:22]: We need to add new env vars for testing to `pytest.ini.example`

Comment (digimaun) [reply azext_iot/tests/central/__init__.py:22]: As well as CI definitions/scripts.

**Comment by digimaun** [comment]
When including local test execution results for PR validation, please don't skip tests.

File: HISTORY.rst [hunk 34]
```diff
@@ -3,6 +3,19 @@
 Release History
 ===============
 
+0.11.0
++++++++++++++++
+
+**IoT Central updates**
+
+* Adds preview commands (v1.1-preview).
```
Comment (digimaun) [inline HISTORY.rst:11]: Can we itemize the command groups that were added?

File: HISTORY.rst [hunk 35]
```diff
@@ -3,6 +3,19 @@
...
+* Adds x-ms-client-request-id header for each request
+
+**Breaking Changes**
+
+* List commands like `az iot central device list` and others,
```
Comment (digimaun) [inline HISTORY.rst:16]: Can we be specific on which commands have breaking changes? Are these GA (non-preview) commands?

File: dev_requirements (hunk 21 above)
Comment (lucadruda) [reply dev_requirements:16]: correct. I fixed it

File: jobdata.json (hunk 22 above)
Comment (lucadruda) [reply jobdata.json:1]: good catch

File: azext_iot/central/models/v1/role.py (hunk 23 above)
Comment (lucadruda) [reply azext_iot/central/models/v1/role.py:12]: removed everywhere

File: azext_iot/central/params.py (hunk 28 above)
Comment (lucadruda) [reply azext_iot/central/params.py:292]: it was actually meant in that way. orgs are only available for 1.1-preview. Same for others.

File: azext_iot/central/_help.py (hunk 30 above)
Comment (lucadruda) [reply azext_iot/central/_help.py:567]: we prefer to be consistent to REST api naming

File: azext_iot/central/_help.py (hunk 32 above)
Comment (lucadruda) [reply azext_iot/central/_help.py:485]: sure

File: azext_iot/central/params.py (hunk 26 above)
Comment (lucadruda) [reply azext_iot/central/params.py:169]: fixed

File: azext_iot/central/params.py (hunk 27 above)
Comment (lucadruda) [reply azext_iot/central/params.py:108]: fixed

File: azext_iot/central/_help.py (hunk 31 above)
Comment (lucadruda) [reply azext_iot/central/_help.py:664]: this is consistent to other commands (e.g. --device-id, --device-name). we'd like to reduce breaking changes at minimum and eventually align all in a different release

File: azext_iot/tests/central/__init__.py (hunk 33 above)
Comment (lucadruda) [reply azext_iot/tests/central/__init__.py:22]: done

File: HISTORY.rst (hunk 35 above)
Comment (lucadruda) [reply HISTORY.rst:16]: added. some of them are GA, some preview. worth specifying in history?

File: azext_iot/tests/central/test_iot_central_devices_int.py [hunk 36]
```diff
@@ -88,7 +104,9 @@ def test_device_connect(self):
         command = "iot central device compute-device-key --pk {} -d {}".format(
```
Comment (digimaun) [inline azext_iot/tests/central/test_iot_central_devices_int.py:104]: Would be **great** to eliminate the need for azext_iot_central_primarykey and azext_iot_central_scope_id simplifying test setup. This `test_device_connect` could create a new device and grab the key for it and scope ID from the API. Thoughts?

In general, we need to keep an eye out for other test setup simplifications.

File: HISTORY.rst (hunk 35 above)
Comment (digimaun) [reply HISTORY.rst:19]: The history content looks much nicer now. Just formatting appears a bit off in the first list

image

Breaking changes on GA commands are concerning though, has the impact been looked into? Are customers to be notified?

File: azext_iot/tests/central/test_iot_central_devices_int.py (hunk 36 above)
Comment (lucadruda) [reply azext_iot/tests/central/test_iot_central_devices_int.py:115]: we have open work items for the next sprint. they'll cover a lot of code refactoring and simplifications

File: HISTORY.rst (hunk 34 above)
Comment (lucadruda) [reply HISTORY.rst]: added

File: HISTORY.rst (hunk 35 above)
Comment (lucadruda) [reply HISTORY.rst:20]: fixed. list commands match GA apis but they were not present in CLI before so no customer impact

File: azext_iot/tests/central/__init__.py [hunk 37]
```diff
@@ -148,49 +185,52 @@ def _create_device_template(self):
...
             try:
-                self.cmd(command, checks=[self.check("result", "success")])
+                self.cmd(
+                    command,
+                    api_version=api_version,
+                    checks=[self.check("result", "success")],
+                )
                 return
             except:
                 time.sleep(10)
```
Comment (valluriraj) [inline azext_iot/tests/central/__init__.py:235]: with the current logic if retry fails after n attempts user does not get indication about failure

**Review by valluriraj — APPROVED** [review]

File: azext_iot/tests/central/__init__.py (hunk 37 above)
Comment (lucadruda) [reply azext_iot/tests/central/__init__.py:238]: fixed

**Review by vilit1 — APPROVED** [review]

**Pull request merged**